algorithms/C/linked-lists/glthreads.c

```diff
@@ -0,0 +1,133 @@
+pointer and a pointer to the data, so the structure of the node will contain
+3 data members(next, prev, data). In Glue based doubly linked-list there will
+only be 2 data members left and right pointers. As for the data, it will be 
+glued on top of the node.
+
+Illustration
+-------------
+Traditional doubly linked-list:
+                         _____________                    _____________
+                    ___ | application |               ___| application |
+                   |    |   data      |              |   |    data     |
+                   |    |_____________|              |   |_____________|
+         __________|_________              __________|_________
+        | prev | data | next |____________| prev | data | next |
+        |______|______|______|            |______|______|______|
+
+
+Glue based doubly linked-list:
+        
+         _____________             _____________
+        | application |           | application |
+        |    data     |           |    data     |
+        |_____________|           |_____________|
+        | prev | next |___________| prev | next |
+        |______|______|           |______|______|
+
+*/
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#define HEAD(list) (list->head)
+#define OFFSETOF(structure, field) (long)&((structure*)0)->field
+#define ITERATE_GL_THREADS_BEGIN(lstptr, struct_type, ptr)      \
+{                                                               \
+    glthread_node *_glnode = NULL, *_next = NULL;               \
+    for(_glnode = lstptr->head; _glnode; _glnode = _next){      \
+        _next = _glnode->next;                                  \
+        ptr = (struct_type *)((char *)_glnode - lstptr->offset);
+#define ITERATE_GL_THREADS_ENDS }}
+
+typedef struct glthread_node_{
+    struct glthread_node_* next;
+    struct glthread_node_* prev;
+}glthread_node;
+
+typedef struct{
+   glthread_node* head;
+   unsigned int offset;
+}gldll;
+
+void init_glthread_node(glthread_node* node){
+    node->next = NULL;
+    node->prev = NULL;
+}
+
+void init_glthread_list(gldll* list, unsigned int offset){
+    list->head = NULL;
+    list->offset = offset;
+}
+
+void glthread_pushfront(gldll* list, glthread_node* node){
+    glthread_node* cur = HEAD(list);
+    if(cur == NULL){
+        HEAD(list) = node;
+        return;
+    }
+    cur->prev = node;
+    node->next = cur;
+    HEAD(list) = node;
+}
+
+typedef struct{
+    char name[30];
+    int age, weight, height;
+    glthread_node glnode;
+}person;
+
+void print(person* per){
+    printf("Name: %s\n", per->name);
+    printf("Height: %i\n", per->height);
+    printf("Age: %i\n", per->age);
+    printf("Weight: %i\n", per->weight);
+	printf("------------\n");
+}
+
+int main(){
+    gldll* list = malloc(sizeof(*list));
+
+    person* rahul = malloc(sizeof(*rahul));
+    person* rohit = malloc(sizeof(*rohit));
+
+    strcpy(rahul->name, "rahul");
```

Comment:
On allocation failure `malloc ()` returns a `NULL` pointer. So in that case, all the attempts to dereference `rahul`, `rohit` or `list` would lead to undefined behaviour.

Reply:
i'll make the changes